// ngx-vz-cell/lib/code/output/visualizations/value/value.component.ts
import { BaseVisualizationComponent } from '../baseVisualization.component';
import { Options } from './../../../../models/options.model';
import { ExecutionOutput } from '../../../../models/output.model';
import {Component} from "@angular/core";
import {Value} from "./value.model";

@Component({
    moduleId: module.id,
    styleUrls: ["value.component.css"],
    templateUrl: "value.component.html",
})
export class ValueVisualizationComponent extends BaseVisualizationComponent {
    private value: Value;

    constructor() {
        super();
    }

    // Sets the output to be rendered by this component.
    public setOutput(output: ExecutionOutput): void {

        // Get results from output.
        let results = output.getResults();

        if (results) {
            let content = results.getContent();

            if (content) {
                this.value = new Value(content);
            }
        }
    }

    // Sets the options associated with this component.
    public setOptions(options: Options): void {
        // For now, no options used in this component.
    }
}
