// ngx-vz-cell/lib/code/output/visualizations/chart/graph/data.ts
/**
 * Copyright 2017 Verizon Laboratories. All rights reserved.
 * See provided LICENSE file for use of this source code.
 */

// Data implements the type returned by NorthStar services as execution results.
export class Data {
    private columns: string[];
    private rows: string[];

    constructor(data: any) {
        this.columns = new Array<string>();
        this.rows = new Array<string>();

        // Note that we are expecting data to have the format
        //  {
        //      "type": string,
        //      "content": {
        //          "columns": array,
        //          "rows": array,
        //      }
        //  }
        // See NorthStar confluence pages for more information.
        // (https://confluence.verizon.com/display/NPDNOR/Portal)
        //
        if (data && data.content) {
            this.columns = data.content.columns;
            this.rows = data.content.rows;
        }
    }

    // Returns the data columns.
    public getColumns(): string[] {
        return this.columns;
    }

    // Returns the index of the column with the specified name.
    public getColumnIndex(name: string): number {
        return this.columns.indexOf(name);
    }

    // Sets the data columns.
    public setColumns(columns: string[]): void {
        this.columns = columns;
    }

    // Returns the data rows.
    public getRows(): string[] {
        return this.rows;
    }

    // Sets the data rows.
    public setRows(rows: string[]): void {
        this.rows = rows;
    }
}
