// ngx-vz-cell/lib/code/output/visualizations/chart/graph/multibar.graph.ts
/**
 * Copyright 2017 Verizon Laboratories. All rights reserved.
 * See provided LICENSE file for use of this source code.
 */
import {BaseGraph} from "./graph";
import {Data} from "./data";
import {GraphSettings} from "./settings";
import {GraphType} from "./type";
import {isNumeric} from "rxjs/util/isNumeric";

declare let d3: any;

export class MultiBarGraph extends BaseGraph {
    constructor(data, graphSettings: GraphSettings) {
        super(data, graphSettings);
    }

    // Returns type of this graph.
    public getType(): string {
        return GraphType.MultiBar;
    }

    // Implement base class abstract method.
    public isChart(): boolean {
        return true;
    }

    // getGraphData returns the data in the format expected by d3 library.
    protected getGraphData(data: Data, graphSettings: GraphSettings): any {
        let bars = [];
        let rows = data.getRows();
        let selectedColumns = graphSettings.getSelectedColumns();

        for (let selectedColumn of selectedColumns) {
            let columnIndex = data.getColumnIndex(selectedColumn);

            // Translate row values into array of points. Note that we are expecting rows
            // to be in the following format:
            // {
            //      "rows": [
            //          [value1.1, value2.1],
            //          [value1.2, value2.2],
            //          ...
            //      ]
            //  }
            let points = [];
            let xValue: number = 0;
            for (let row of rows) {

                // if the column contains invalid data. Skip it.
                if (isNumeric(row[columnIndex]) === false) {
                    continue;
                }

                let point = {
                    x: xValue,
                    y: row[columnIndex],
                };

                points.push(point);
            }

            if (points.length > 0) {
                let bar = {
                    key: selectedColumn,
                    values: points,
                };

                bars.push(bar);
            }
        }

        return bars;
    }

    // getDefaultOptions returns options as expected by d3 library.
    protected getGraphOptions(): any {
        const ChartOptions = {
            chart: {
                type: GraphType.MultiBar,
                height: BaseGraph.DefaultHeight,
                margin: BaseGraph.DefaultMargin,
                clipEdge: true,
                duration: 500,
                stacked: true,
                reduceXTicks: true,
                showControls: true,
                groupSpacing: 0.1,
                xAxis: {
                    axisLabel: BaseGraph.DefaultXAxisLabel,
                    showMaxMin: true,
                },
                yAxis: {
                    axisLabel: BaseGraph.DefaultYAxisLabel,
                    axisLabelDistance: BaseGraph.DefaultYAxisLabelDistance,
                },
            },
        };
        return ChartOptions;
    }
}
