// ngx-vz-cell/lib/code/output/visualizations/chart/graph/settings.ts
/**
 * Copyright 2017 Verizon Laboratories. All rights reserved.
 * See provided LICENSE file for use of this source code.
 */
import {DefaultGraphType} from "./type";

export class GraphSettings {
    public type: string;
    public selectedColumns: string[];
    public chartOptions: any;

    constructor(obj?: Object) {
        this.type = DefaultGraphType;
        this.selectedColumns = [];
        this.chartOptions = {};

        if (obj) {
            this.unmarshal(obj as GraphSettings);
        }
    }

    // Returns graph type.
    public getType(): string {
        return this.type;
    }

    // Sets the graph type.
    public setType(type: string): void {
        this.type = type;
    }

    // Returns selected columns.
    public getSelectedColumns(): string[] {
        return this.selectedColumns;
    }

    // Sets selected columns.
    public setSelectedColumns(selectedColumns: string[]): void {
        this.selectedColumns = selectedColumns;
    }

    // Returns chart options.
    public getChartOptions(): any {
        return this.chartOptions;
    }

    private unmarshal(obj: GraphSettings) {
        if (obj.type) {
            this.type = obj.type;
        }

        if (obj.selectedColumns) {
            this.selectedColumns = obj.selectedColumns;
        }

        if (obj.chartOptions) {
            this.chartOptions = obj.chartOptions;
        }
    }
}
