// ngx-vz-cell/lib/models/options.model.ts
import {VisualizationSettings} from "./visualizationSettings.model";
export class Options {
    public showCode: boolean;
    public showOutput: boolean;
    public showConfiguration: boolean;
    public timeout: number;
    public mainFunction: string;
    public visualization: VisualizationSettings;
    public status: string;

    constructor(obj?: Object) {
        if (obj) {
            this.unmarshal(obj as Options);
        } else {
            this.showCode = true;
            this.showOutput = true;
            this.visualization = new VisualizationSettings();
            this.mainFunction = "main";
            this.timeout = 180;
            this.showConfiguration = true;
        }
    }

    private unmarshal(obj: Options) {
        this.showCode = obj.showCode;
        this.showOutput = obj.showOutput;
        this.timeout = obj.timeout;
        this.mainFunction = obj.mainFunction;
        this.visualization = new VisualizationSettings(obj.visualization);
        this.status = obj.status;
        this.showConfiguration = obj.showConfiguration;
    }
}
