// ngx-vz-cell/lib/code/code.module.ts
/*General*/
import {BootstrapModalModule} from "angular2-modal/plugins/bootstrap";
import {NgModule} from "@angular/core";
import {CommonModule} from "@angular/common";
import {FormsModule} from "@angular/forms";
import {ModalModule} from "angular2-modal";
/*Cell*/
import {CellOptionsModal} from "./options/options.component";
import {CellInputModal} from "./inputs/inputs.component";
import {CodemirrorModule} from "ng2-codemirror";
import {StatusIndicatorComponent} from "./status/status.component";
import {InlineEdit} from "../inlineEdit/inlineEdit.component";
import {Output} from "./output/output.component";

/*Visualizations*/
import {VisualizationComponent} from "./output/visualizations/visualizations.component";
import {VisualizationModule} from "./output/visualizations/visualizations.module";
import {CodeCellComponent} from "./code.component";

@NgModule({
    declarations: [
        CellOptionsModal,
        CellInputModal,
        VisualizationComponent,
        StatusIndicatorComponent,
        CodeCellComponent,
        InlineEdit,
        Output,
    ],
    entryComponents: [
        CellOptionsModal,
        CellInputModal,
    ],
    exports: [
        CodeCellComponent,
        InlineEdit,
        Output,
    ],
    imports: [
        CommonModule,
        FormsModule,
        ModalModule.forRoot(),
        BootstrapModalModule,
        VisualizationModule,
        CodemirrorModule,
    ],
    providers: [],
})
export class CodeModule {

}
